public/sw.js

```diff
@@ -3,20 +3,61 @@
 /* eslint-disable no-unused-vars */
 /* eslint-disable no-undef */
 /* eslint-disable no-restricted-globals */
-const version = 'v0.3.0';
+/* eslint-disable no-console */
+
+const version = 'v0.3.1';
 const cacheName = 'simorghCache_v1';
 
-const service = self.location.pathname.split('/')[1];
-const hasOfflinePageFunctionality = false;
-const OFFLINE_PAGE = `/${service}/offline`;
+// Track PWA clients
+const pwaClients = new Map();
 
-self.addEventListener('install', event => {
-  event.waitUntil(async () => {
-    const cache = await caches.open(cacheName);
-    if (hasOfflinePageFunctionality) await cache.add(OFFLINE_PAGE);
-  });
-});
+console.log(`[SW v${version}] Service Worker loaded.`);
+
+// --------------------
+// Helper Functions
+// --------------------
+const getServiceFromUrl = url => new URL(url).pathname.split('/')[1];
+const getOfflinePageUrl = service => `/${service}/offline`;
+
+const cacheResource = async (cache, url) => {
+  try {
+    const response = await fetch(url);
+    if (response.ok) await cache.put(url, response.clone());
+    return response;
+  } catch (err) {
+    console.error(`[SW v${version}] Failed to cache ${url}:`, err);
+    return null;
+  }
+};
 
+const cacheOfflinePageAndResources = async service => {
+  const cache = await caches.open(cacheName);
+  const offlinePageUrl = new URL(
+    getOfflinePageUrl(service),
+    self.location.origin,
+  ).href;
+  if (await cache.match(offlinePageUrl)) return;
+
+  const resp = await cacheResource(cache, offlinePageUrl);
+  if (!resp || !resp.ok) return;
+
+  console.log(`[SW v${version}] Cached offline page for ${service}`);
+
+  const html = await resp.text();
+  const scriptSrcs = [
+    ...html.matchAll(/<script[^>]+src=["']([^"']+)["']/g),
+  ].map(m => m[1]);
+  const linkHrefs = [...html.matchAll(/<link[^>]+href=["']([^"']+)["']/g)].map(
+    m => m[1],
+  );
+  const resources = [...scriptSrcs, ...linkHrefs]
+    .filter(url => url.startsWith('/') || url.startsWith(self.location.origin))
+    .map(url => new URL(url, self.location.origin).href);
```

Comment:
Could this have been achieved using the queryselector api rather than regex? You could parse the doc as html using https://developer.mozilla.org/en-US/docs/Web/API/DOMParser 

Then found the link element using query selector

Might be a bit less error prone/cleaner?